### PR TITLE
wireshark: build with QT 6

### PR DESCRIPTION
**Pull request by lorenz**
###### Description of changes

Wireshark 4 supports Qt 6 so let's use that.

Tested on GNOME 44 Wayland, nothing was broken and the window decorations now look better.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/sniffers/wireshark/default.nix [hunk 1]
```diff
@@ -108,11 +109,11 @@ stdenv.mkDerivation {
...
     glib
     zlib
-  ] ++ lib.optionals withQt (with qt5; [ qtbase qtmultimedia qtsvg qttools ])
-  ++ lib.optionals (withQt && stdenv.isLinux) [ qt5.qtwayland ]
+  ] ++ lib.optionals withQt (with qt6; [ qtbase qtmultimedia qtsvg qttools qt5compat ])
+  ++ lib.optionals (withQt && stdenv.isLinux) [ qt6.qtwayland ]
   ++ lib.optionals stdenv.isLinux [ libcap libnl sbc ]
   ++ lib.optionals stdenv.isDarwin [ SystemConfiguration ApplicationServices gmp ]
-  ++ lib.optionals (withQt && stdenv.isDarwin) (with qt5; [ qtmacextras ]);
+  ++ lib.optionals (withQt && stdenv.isDarwin) (with qt6; [ qtmacextras ]);
```
Comment (teutat3s) [inline pkgs/applications/networking/sniffers/wireshark/default.nix:116]: Looks like `qtmacextras` does not exist as a module in Qt6. Does this break WireShark on macOS using Qt6 or is this no longer required with the new version?

**Review by teutat3s — CHANGES_REQUESTED** [review]

File: pkgs/applications/networking/sniffers/wireshark/default.nix (hunk 1 above)
Comment (Tungsten842) [reply pkgs/applications/networking/sniffers/wireshark/default.nix:116]: > Looks like `qtmacextras` does not exist as a module in Qt6. Does this break WireShark on macOS using Qt6 or is this no longer required with the new version?

Qtmacextra [no longer exists](https://www.qt.io/blog/qt-extras-modules-in-qt-6) in qt6, this line should be deleted.

Comment (lorenz) [reply pkgs/applications/networking/sniffers/wireshark/default.nix]: Removed that line. Sadly I have no macOS to test the Darwin build on.

**Comment by yu-re-ka** [comment]
Builds and runs fine on aarch64-darwin

**Comment by yu-re-ka** [comment]
CC @NixOS/darwin-maintainers is this a regression for x86_64-darwin and if yes how do we fix it?

**Comment by teutat3s** [comment]
Can reproduce the same failure on `x86_64-darwin`. Tried building wireshark from `nixos-unstable` on the same machine and it fails with a different error.

```
FAILED: capture/CMakeFiles/caputils.dir/capture-pcap-util.c.o
```

It looks like it did still build fine here https://github.com/NixOS/nixpkgs/pull/227852

**Comment by Tungsten842** [comment]
> CC @NixOS/darwin-maintainers is this a regression for x86_64-darwin and if yes how do we fix it?

I don't think that this is a regression, it was already broken here: https://hydra.nixos.org/eval/1794647?filter=wireshark&compare=1794631&full=#tabs-still-fail

**Comment by wegank** [comment]
```
20 packages built:
compactor credslayer credslayer.dist dbmonster haka hfinger hfinger.dist ostinato python310Packages.pyshark python310Packages.pyshark.dist python311Packages.pyshark python311Packages.pyshark.dist qtwirediff termshark tshark tshark.dev wifite2 wifite2.dist wireshark wireshark.dev
```

**Review by wegank — APPROVED** [review]

**Comment by quentinmit** [comment]
I just found this PR after I updated to 23.05; I was successfully using Wireshark on x86_64-darwin before this PR, but it now fails to build with an error in Qt 6. For now I'm just using an overlay to revert to qt5 - would you maybe consider having the package work with both?

**Comment by lorenz** [comment]
I don't think this is related. Wireshark was broken on x86_64-darwin before this PR (see comment above), also the error message (on Hydra, don't have a macOS handy to test) indicates that it is likely related to a macOS SDK issue (see also https://github.com/NixOS/nixpkgs/issues/210558).,

**Comment by quentinmit** [comment]
> I don't think this is related. Wireshark was broken on x86_64-darwin before this PR (see comment above), also the error message (on Hydra, don't have a macOS handy to test) indicates that it is likely related to a macOS SDK issue (see also #210558).,

@lorenz I'm not sure what was happening on Hydra, but Wireshark built for me before this PR, and it builds again for me with this overlay:

```
  wireshark-qt5 = (prev.wireshark.overrideAttrs (old: {
    pname = "wireshark-qt5";

    cmakeFlags = prev.lib.lists.remove "-DUSE_qt6=ON" old.cmakeFlags ++ [
      "-DCMAKE_BUILD_WITH_INSTALL_NAME_DIR=ON"
    ];
    preFixup = old.preFixup + final.lib.optionalString final.stdenv.isDarwin ''
      # https://bugreports.qt.io/browse/QTBUG-81370
      qtWrapperArgs+=(--set QT_MAC_WANTS_LAYER 1)
      # Remove the executable bit from plugins so that Nix doesn't try to wrap them
      find $out/Applications/Wireshark.app/Contents/PlugIns ! -type d -executable -exec chmod a-x {} \;
    '';
  })).override {
    qt6 = final.qt5 // {
      qt5compat = null;
    };
    stdenv = if final.stdenv.isDarwin then final.darwin.apple_sdk_11_0.stdenv else final.stdenv;
    buildPackages = final.buildPackages // final.lib.optionalAttrs final.stdenv.isDarwin {
      stdenv = final.buildPackages.darwin.apple_sdk_11_0.stdenv;
    };
  };
```

Note that most of this is fixing unrelated runtime problems (the app still builds and runs without the `cmakeFlags` and `preFixup`, you just can't load all the plugins and there are graphical glitches).

**Pull request merged**